src/services/domain/roles/roles.service.ts

```diff
@@ -89,96 +82,193 @@ export class RolesService {
-        if ('hubID' in parent) {
-          const hubResult = await this.ensureHubRolesResult(
-            hubsMap,
-            (parent as ICommunity).hubID,
+      switch (credential.type) {
+        case AuthorizationCredential.ORGANIZATION_MEMBER:
+          await this.addOrganizationMemberRole(
+            orgsMap,
+            credential,
             contributorID
           );
-          const groupResult = new RolesResult(group.name, group.id, group.name);
-          hubResult.userGroups.push(groupResult);
-        } else {
-          const orgResult = await this.ensureOrganizationRolesResult(
-            orgsMap,
-            (parent as IOrganization).id,
+          break;
+        case AuthorizationCredential.HUB_MEMBER:
+          await this.addHubMemberRole(hubsMap, credential, contributorID);
+          break;
+        case AuthorizationCredential.HUB_HOST:
+          await this.addHubHostRole(hubsMap, credential, contributorID);
+          break;
+        case AuthorizationCredential.CHALLENGE_LEAD:
+          await this.addChallengeLeadRole(hubsMap, credential, contributorID);
+          break;
+        case AuthorizationCredential.CHALLENGE_MEMBER:
+          await this.addChallengeMemberRole(hubsMap, credential, contributorID);
+          break;
+        case AuthorizationCredential.OPPORTUNITY_MEMBER:
+          await this.addOpportunityMemberRole(
+            hubsMap,
+            credential,
             contributorID
           );
-          const groupResult = new RolesResult(group.name, group.id, group.name);
-          orgResult.userGroups.push(groupResult);
-        }
+          break;
+        case AuthorizationCredential.OPPORTUNITY_LEAD:
+          await this.addOpportunityLeadRole(hubsMap, credential, contributorID);
+          break;
+        case AuthorizationCredential.USER_GROUP_MEMBER:
+          await this.addUserGroupMemberRole(
+            credential,
+            hubsMap,
+            contributorID,
+            orgsMap
+          );
+          break;
       }
     }
+  }
 
-    for (const hubResult of hubsMap.values()) {
-      membership.hubs.push(hubResult);
-    }
-    for (const orgResult of orgsMap.values()) {
-      membership.organizations.push(orgResult);
+  private async addUserGroupMemberRole(
+    credential: ICredential,
+    hubsMap: Map<string, RolesResultHub>,
+    contributorID: string,
+    orgsMap: Map<string, RolesResultOrganization>
+  ) {
+    const group = await this.userGroupService.getUserGroupOrFail(
+      credential.resourceID
+    );
+    const parent = await this.userGroupService.getParent(group);
+    if ('hubID' in parent) {
+      const hubResult = await this.ensureHubRolesResult(
+        hubsMap,
+        (parent as ICommunity).hubID,
+        contributorID
+      );
+      const groupResult = new RolesResult(group.name, group.id, group.name);
+      hubResult.userGroups.push(groupResult);
+    } else {
```

Comment:
Since we don't know what implements `IGroupable` you must have an explicit check `isIOrganization` and also infer the type.